crates/db_views_actor/src/community_follower_view.rs

```diff
@@ -88,4 +97,185 @@ impl CommunityFollowerView {
       .load::<CommunityFollowerView>(conn)
       .await
   }
+
+  pub async fn list_approval_required(
+    pool: &mut DbPool<'_>,
+    community_id: CommunityId,
+    pending_only: bool,
+    page: Option<i64>,
+    limit: Option<i64>,
+  ) -> Result<Vec<PendingFollow>, Error> {
+    let conn = &mut get_conn(pool).await?;
+    let (limit, offset) = limit_and_offset(page, limit)?;
+    let person_alias_1 = diesel::alias!(person as person1);
+
+    // check if the community already has an accepted follower from the same instance
+    let is_new_instance = not(exists(
+      community_follower::table.filter(
+        community_follower::person_id
+          .eq(person_alias_1.field(person::id))
+          .and(person::instance_id.eq(person_alias_1.field(person::instance_id)))
+          .and(community_follower::state.eq(CommunityFollowerState::Accepted)),
+      ),
+    ));
```

Comment:
This fails to compile, but the diesel error is very unclear to me. Any suggestions how to fix it?

```

error[E0277]: the trait bound `FromClause<query_source::joins::Join<lemmy_db_schema::schema::community_follower::table, JoinOn<query_source::joins::Join<JoinOn<query_source::joins::Join<lemmy_db_schema::schema::community_follower::table, lemmy_db_schema::schema::person::table, Inner>, diesel::expression::grouped::Grouped<diesel::expression::operators::Eq<NullableExpression<lemmy_db_schema::schema::community_follower::person_id>, NullableExpression<lemmy_db_schema::schema::person::id>>>>, Alias<person1>, Inner>, diesel::expression::grouped::Grouped<diesel::expression::operators::Eq<lemmy_db_schema::schema::community_follower::person_id, AliasedField<person1, lemmy_db_schema::schema::person::id>>>>, Inner>>: query_builder::from_clause::AsQuerySource` is not satisfied
--> crates/db_views_actor/src/community_follower_view.rs:133:8
|
133 |       .select((person::all_columns, is_new_instance))
|        ^^^^^^ the trait `query_builder::from_clause::AsQuerySource` is not implemented for `FromClause<Join<table, JoinOn<Join<JoinOn<Join<table, table, Inner>, Grouped<Eq<Nullable<...>, ...>>>, ..., ...>, ...>, ...>>`, which is required by `BoxedSelectStatement<'_, ((Integer, Integer, Timestamptz, lemmy_db_schema::schema::sql_types::CommunityFollowerState, Nullable<Integer>), (Integer, Text, Nullable<Text>, Nullable<Text>, diesel::sql_types::Bool, Timestamptz, Nullable<Timestamptz>, Text, Nullable<Text>, diesel::sql_types::Bool, Nullable<Text>, Text, Timestamptz, Nullable<Text>, diesel::sql_types::Bool, Text, Nullable<Text>, Nullable<Text>, diesel::sql_types::Bool, Nullable<Timestamptz>, Integer), (Integer, Text, Nullable<Text>, Nullable<Text>, diesel::sql_types::Bool, Timestamptz, Nullable<Timestamptz>, Text, Nullable<Text>, diesel::sql_types::Bool, Nullable<Text>, Text, Timestamptz, Nullable<Text>, diesel::sql_types::Bool, Text, Nullable<Text>, Nullable<Text>, diesel::sql_types::Bool, Nullable<Timestamptz>, Integer)), FromClause<JoinOn<query_source::joins::Join<JoinOn<query_source::joins::Join<lemmy_db_schema::schema::community_follower::table, lemmy_db_schema::schema::person::table, Inner>, diesel::expression::grouped::Grouped<diesel::expression::operators::Eq<NullableExpression<lemmy_db_schema::schema::community_follower::person_id>, NullableExpression<lemmy_db_schema::schema::person::id>>>>, Alias<person1>, Inner>, diesel::expression::grouped::Grouped<diesel::expression::operators::Eq<lemmy_db_schema::schema::community_follower::person_id, AliasedField<person1, lemmy_db_schema::schema::person::id>>>>>, Pg>: SelectDsl<_>`
...

error[E0271]: type mismatch resolving `<Join<table, JoinOn<Join<JoinOn<Join<table, table, Inner>, Grouped<Eq<..., ...>>>, ..., ...>, ...>, ...> as AppearsInFromClause<...>>::Count == Once`
--> crates/db_views_actor/src/community_follower_view.rs:133:8
|
133 |       .select((person::all_columns, is_new_instance))
|        ^^^^^^ expected `MoreThanOnce`, found `Once`
|
```

Reply:
I'll take a look.

Reply:
Okay, these are two separate queries that seem to be unecessarily trying to be joined (that are replicating a lot of their own joins, except for community.

Why do you need `is_new_instance`, why not just select the `community_follower::state` ? It looks like you are doing the `check_has_followers_from_instance` as a separate thing in the apub code anyway, so for the `get_pending_follows` endpoint, it makes sense to just return the `community_followore::state`.

Reply:
The reason for is_new_instance is explained [here in the rfc](https://github.com/LemmyNet/rfcs/blob/main/0005-private-communities.md#frontends-and-apps). Basically to warn mods that private community content could be leaked by a new instance, eg if it is malicious or doesnt properly handle private content.

You can check the previous commit which has a working implementation, but using a separate sql query for each item which is inefficient. So its a matter of merging these two queries into one.

Reply:
You *might* be able to do something with a group by, but its not really clean. What I'd do, is on the approvals page, show that warning regardless in the front end (IE something like *be aware that admins of federated users you approve, will be able to access the content of your private community*).

Then just return `(Person, community_follower::state)`

Reply:
Then the warning would be completely pointless. If the community already has a follower from lemmy.world, and a new follower from lemmy.world applies, there is no need for any warning. In that case the instance admin has access to the community content already.

Reply:
The endpoint returns pendings from many instances tho, so it'd be fine to have a top-level warning. You don't need to return the community_follower::state if you don't need it I suppose, although it seems useful.

Reply:
Right Ive added the subscribed type. Still need to fix the query.

Reply:
Got it working by writing the query in plain sql first.